src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.cpp

```diff
@@ -12,99 +12,197 @@
-        z_33_factor * real(time_factor * l_3_factor);
-  }
+  const auto& z_22_harmonic = cached_z_harmonics(l_max, 2_st, Spin);
+  const auto& z_33_harmonic = cached_z_harmonics(l_max, 3_st, Spin);
+  bondi_quantity->data() = z_22_harmonic * real(time_factor * l_2_factor) +
+                           z_33_harmonic * real(time_factor * l_3_factor);
 }
 
 template <int Spin>
-void LinearizedBondiSachs::assign_du_components_from_l_factors(
+void assign_du_components_from_l_factors(
     const gsl::not_null<SpinWeighted<ComplexDataVector, Spin>*>
         du_bondi_quantity,
     const std::complex<double>& l_2_factor,
     const std::complex<double>& l_3_factor, const size_t l_max,
-    const double time) const noexcept {
+    const double frequency, const double time) noexcept {
   const std::complex<double> time_factor =
-      frequency_ * (std::complex<double>(0.0, 1.0) * cos(frequency_ * time) +
-                    sin(frequency_ * time));
+      frequency * (std::complex<double>(0.0, 1.0) * cos(frequency * time) -
+                   sin(frequency * time));
 
   // mode constants for the computation of the linearized solution.
-  Spectral::Swsh::SpinWeightedSphericalHarmonic y_22{Spin, 2_st, 2};
-  Spectral::Swsh::SpinWeightedSphericalHarmonic y_2m2{Spin, 2_st, -2};
-  Spectral::Swsh::SpinWeightedSphericalHarmonic y_33{Spin, 3_st, 3};
-  Spectral::Swsh::SpinWeightedSphericalHarmonic y_3m3{Spin, 3_st, -3};
-
-  const auto& collocation_metadata =
-      Spectral::Swsh::cached_collocation_metadata<
-          Spectral::Swsh::ComplexRepresentation::Interleaved>(l_max);
-  for (const auto collocation_point : collocation_metadata) {
-    // assign to collocation points
-    const std::complex<double> z_22_factor =
-        (y_22.evaluate(collocation_point.theta, collocation_point.phi) +
-         y_2m2.evaluate(collocation_point.theta, collocation_point.phi));
-    const std::complex<double> z_33_factor =
-        (y_33.evaluate(collocation_point.theta, collocation_point.phi) -
-         y_3m3.evaluate(collocation_point.theta, collocation_point.phi));
-    du_bondi_quantity->data()[collocation_point.offset] =
-        z_22_factor * real(time_factor * l_2_factor) +
-        z_33_factor * real(time_factor * l_3_factor);
-  }
+  const auto& z_22_harmonic = cached_z_harmonics(l_max, 2_st, Spin);
+  const auto& z_33_harmonic = cached_z_harmonics(l_max, 3_st, Spin);
+  du_bondi_quantity->data() = z_22_harmonic * real(time_factor * l_2_factor) +
+                              z_33_harmonic * real(time_factor * l_3_factor);
+}
+}  // namespace LinearizedBondiSachs_detail
+
+LinearizedBondiSachs::LinearizedBondiSachs(
+    const std::vector<std::complex<double>>& mode_constants,
+    const double extraction_radius, const double frequency) noexcept
+    : SphericalMetricData{extraction_radius}, frequency_{frequency} {
+  auto mode_constant_copy = mode_constants;
+  mode_constant_copy.resize(2, std::complex<double>(0.0, 0.0));
```

Comment:
What does this resizing do? Is the intention that the user may specify only one mode, and the other defaults to zero? In that case I suggest you just remove the default behavior, which means you can switch the `std::vector` to a `std::array`. If I misunderstood this, I suggest to add a comment what's going on :)